File: HikvisionService/Controllers/CameraController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using HikvisionService.Services;
using HikvisionService.Data;
using HikvisionService.Models;

namespace HikvisionService.Controllers;

[ApiController]
[Route("api/[controller]")]
public class CameraController : ControllerBase
{
    private readonly IHikvisionService _hikvisionService;
    private readonly ILogger<CameraController> _logger;
    private readonly HikvisionDbContext _context;

    public CameraController(IHikvisionService hikvisionService, ILogger<CameraController> logger, HikvisionDbContext context)
    {
        _hikvisionService = hikvisionService;
        _logger = logger;
        _context = context;
    }

    /// <summary>
    /// Get all cameras
    /// </summary>
    [HttpGet]
    public async Task<IActionResult> GetCameras()
    {
        try
        {
            var cameras = await _hikvisionService.GetAllCamerasAsync();
            return Ok(cameras);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting cameras");
            return StatusCode(500, "Internal server error");
        }
    }

    /// <summary>
    /// Get camera by ID
    /// </summary>
    [HttpGet("{id}")]
    public async Task<IActionResult> GetCamera(long id)
    {
        try
        {
            var camera = await _hikvisionService.GetCameraByIdAsync(id);
            if (camera == null)
            {
                return NotFound($"Camera with ID {id} not found");
            }
            return Ok(camera);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting camera {CameraId}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    /// <summary>
    /// Get available files from a camera
    /// </summary>
    [HttpGet("{id}/files")]
    public async Task<IActionResult> GetAvailableFiles(
        long id,
        [FromQuery] DateTime? startTime = null,
        [FromQuery] DateTime? endTime = null,
        [FromQuery] string fileType = "both")
    {
        try
        {
            // Set default time range if not provided (last 4 hours)
            var start = startTime ?? DateTime.Now.AddHours(-4);
            var end = endTime ?? DateTime.Now;

            // Validate file type
            if (fileType != "both" && fileType != "video" && fileType != "photo")
            {
                return BadRequest("FileType must be 'both', 'video', or 'photo'");
            }

            var files = await _hikvisionService.GetAvailableFilesAsync(id, start, end, fileType);
            
            // Transform files to a more API-friendly format
            var response = new
            {
                CameraId = id,
                StartTime = start,
                EndTime = end,
                FileType = fileType,
                TotalFiles = files.Count,
                Files = files.Select(f => new
                {
                    Name = f.Name,
                    Size = f.Size,
                    Date = f.Date,
                    Duration = f.Duration, // Duration in seconds for videos
                    Type = f.Duration > 0 ? "video" : "photo" // Videos have duration, photos don't
                }).ToList()
            };

            return Ok(response);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting available files for camera {CameraId}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    /// <summary>
    /// Test camera connection
    /// </summary>
    [HttpPost("{id}/test-connection")]
    public async Task<IActionResult> TestConnection(long id)
    {
        try
        {
            var isConnected = await _hikvisionService.TestCameraConnectionAsync(id);
            return Ok(new { CameraId = id, IsConnected = isConnected });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error testing connection for camera {CameraId}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    /// <summary>
    /// Create a new camera
    /// </summary>
    [HttpPost]
    public async Task<IActionResult> CreateCamera([FromBody] CreateCameraRequest request)
    {
        try
        {
            // Validate that the store exists
            var storeExists = await _context.Stores.AnyAsync(s => s.Id == request.StoreId);
            if (!storeExists)
            {
                return BadRequest($"Store with ID {request.StoreId} does not exist");
            }

            var camera = new Camera
            {
                StoreId = request.StoreId,
                Name = request.Name,
                IpAddress = request.IpAddress,
                Port = request.Port,
                Username = request.Username,
                Password = request.Password,
                ServerPort = request.ServerPort,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            _context.Cameras.Add(camera);
            await _context.SaveChangesAsync();

            // Load the store for the response
            await _context.Entry(camera)
                .Reference(c => c.Store)
                .LoadAsync();

            _logger.LogInformation("Created camera: {CameraName} with ID: {CameraId}", camera.Name, camera.Id);

            return CreatedAtAction(nameof(GetCamera), new { id = camera.Id }, camera);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating camera");
            return StatusCode(500, "Internal server error");
        }
    }

    /// <summary>
    /// Update existing camera
    /// </summary>
    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateCamera(long id, [FromBody] UpdateCameraRequest request)
    {
        try
        {
            var camera = await _context.Cameras.FindAsync(id);
            if (camera == null)
            {
                return NotFound($"Camera with ID {id} not found");
            }

            // Validate that the store exists
            var storeExists = await _context.Stores.AnyAsync(s => s.Id == request.StoreId);
            if (!storeExists)
            {
                return BadRequest($"Store with ID {request.StoreId} does not exist");
            }

            camera.StoreId = request.StoreId;
            camera.Name = request.Name;
            camera.IpAddress = request.IpAddress;
            camera.Port = request.Port;
            camera.Username = request.Username;
            camera.Password = request.Password;
            camera.ServerPort = request.ServerPort;
            camera.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            // Load the store for the response
            await _context.Entry(camera)
                .Reference(c => c.Store)
                .LoadAsync();

            _logger.LogInformation("Updated camera {CameraId}: {CameraName}", camera.Id, camera.Name);

            return Ok(camera);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating camera {CameraId}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    /// <summary>
    /// Delete camera
    /// </summary>
    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteCamera(long id)
    {
        try
        {
            var camera = await _context.Cameras
                .Include(c => c.FileDownloadJobs)
                .FirstOrDefaultAsync(c => c.Id == id);

            if (camera == null)
            {
                return NotFound($"Camera with ID {id} not found");
            }

            // Delete associated download jobs
            if (camera.FileDownloadJobs.Any())
            {
                _context.FileDownloadJobs.RemoveRange(camera.FileDownloadJobs);
            }

            _context.Cameras.Remove(camera);
            await _context.SaveChangesAsync();

            _logger.LogInformation("Deleted camera {CameraId}: {CameraName}", camera.Id, camera.Name);

            return Ok(new { message = "Camera deleted successfully" });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting camera {CameraId}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    // Request DTOs
    public class CreateCameraRequest
    {
        public long StoreId { get; set; }
        public string Name { get; set; } = string.Empty;
        public string IpAddress { get; set; } = string.Empty;
        public int Port { get; set; } = 554;
        public string? Username { get; set; }
        public string? Password { get; set; }
        public int? ServerPort { get; set; }
    }

    public class UpdateCameraRequest
    {
        public long StoreId { get; set; }
        public string Name { get; set; } = string.Empty;
        public string IpAddress { get; set; } = string.Empty;
        public int Port { get; set; } = 554;
        public string? Username { get; set; }
        public string? Password { get; set; }
        public int? ServerPort { get; set; }
    }
}

File: laravel/app/Http/Controllers/CameraController.php
<?php

namespace App\Http\Controllers;

use App\Models\Camera;
use App\Models\Store;
use App\Services\HikvisionService;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;

class CameraController extends Controller
{
    protected $hikvisionService;

    public function __construct(HikvisionService $hikvisionService)
    {
        $this->hikvisionService = $hikvisionService;
    }

    public function index()
    {
        try {
            $cameras = Camera::with('store')->orderBy('name')->get();
            return view('cameras.index', compact('cameras'));
        } catch (\Exception $ex) {
            Log::error('Error getting cameras', ['error' => $ex->getMessage()]);
            return redirect()->back()->with('error', 'An error occurred while retrieving cameras.');
        }
    }

    public function show($id)
    {
        try {
            $camera = Camera::with('store')->findOrFail($id);
            return view('cameras.show', compact('camera'));
        } catch (\Exception $ex) {
            Log::error("Error getting camera {$id}", ['error' => $ex->getMessage()]);
            return redirect()->route('cameras.index')->with('error', 'Camera not found.');
        }
    }

    public function create()
    {
        $stores = Store::orderBy('name')->get();
        return view('cameras.create', compact('stores'));
    }

    public function store(Request $request)
    {
        $request->validate([
            'store_id' => 'required|exists:stores,id',
            'name' => 'required|string|max:255',
            'ip_address' => 'required|ip',
            'port' => 'required|integer|min:1|max:65535',
            'username' => 'nullable|string|max:100',
            'password' => 'nullable|string|max:100',
            'server_port' => 'nullable|integer|min:1|max:65535',
        ]);

        try {
            Camera::create([
                'store_id' => $request->store_id,
                'name' => $request->name,
                'ip_address' => $request->ip_address,
                'port' => $request->port,
                'username' => $request->username,
                'password' => $request->password,
                'server_port' => $request->server_port,
                'is_online' => false,
            ]);

            Log::info('Created camera', ['name' => $request->name]);
            return redirect()->route('cameras.index')->with('success', 'Camera created successfully.');
        } catch (\Exception $ex) {
            Log::error('Error creating camera', ['error' => $ex->getMessage()]);
            return back()->withInput()->with('error', 'An error occurred while creating the camera.');
        }
    }

    public function edit($id)
    {
        try {
            $camera = Camera::findOrFail($id);
            $stores = Store::orderBy('name')->get();
            return view('cameras.edit', compact('camera', 'stores'));
        } catch (\Exception $ex) {
            Log::error("Error getting camera {$id} for edit", ['error' => $ex->getMessage()]);
            return redirect()->route('cameras.index')->with('error', 'Camera not found.');
        }
    }

    public function update(Request $request, $id)
    {
        $request->validate([
            'store_id' => 'required|exists:stores,id',
            'name' => 'required|string|max:255',
            'ip_address' => 'required|ip',
            'port' => 'required|integer|min:1|max:65535',
            'username' => 'nullable|string|max:100',
            'password' => 'nullable|string|max:100',
            'server_port' => 'nullable|integer|min:1|max:65535',
        ]);

        try {
            $camera = Camera::findOrFail($id);

            $camera->update([
                'store_id' => $request->store_id,
                'name' => $request->name,
                'ip_address' => $request->ip_address,
                'port' => $request->port,
                'username' => $request->username,
                'password' => $request->password,
                'server_port' => $request->server_port,
            ]);

            Log::info('Updated camera', ['id' => $id, 'name' => $request->name]);
            return redirect()->route('cameras.index')->with('success', 'Camera updated successfully.');
        } catch (\Exception $ex) {
            Log::error("Error updating camera {$id}", ['error' => $ex->getMessage()]);
            return back()->withInput()->with('error', 'An error occurred while updating the camera.');
        }
    }

    public function destroy($id)
    {
        try {
            $camera = Camera::with('fileDownloadJobs')->findOrFail($id);

            // Delete associated download jobs
            if ($camera->fileDownloadJobs->count() > 0) {
                $camera->fileDownloadJobs()->delete();
            }

            $cameraName = $camera->name;
            $camera->delete();

            Log::info('Deleted camera', ['id' => $id, 'name' => $cameraName]);
            return redirect()->route('cameras.index')->with('success', 'Camera deleted successfully.');
        } catch (\Exception $ex) {
            Log::error("Error deleting camera {$id}", ['error' => $ex->getMessage()]);
            return redirect()->route('cameras.index')->with('error', 'An error occurred while deleting the camera.');
        }
    }

    public function testConnection($id)
    {
        try {
            $channels = $this->hikvisionService->testCameraConnection($id);
            $message = 'Camera channels are:<br>' . implode('<br>', $channels);
            return redirect()->route('cameras.show', $id)->with('success', $message);
        } catch (\Exception $ex) {
            Log::error("Error testing connection for camera {$id}", ['error' => $ex->getMessage()]);
            return redirect()->route('cameras.show', $id)->with('error', 'An error occurred while testing the camera connection.');
        }
    }
}
